spec/requests/users_controller_spec.rb

```diff
@@ -1,51 +1,2402 @@
 require 'rails_helper'
 
-RSpec.describe UsersController do
+describe UsersController do
   let(:user) { Fabricate(:user) }
 
-  def honeypot_magic(params)
-    get '/u/hp.json'
-    json = JSON.parse(response.body)
-    params[:password_confirmation] = json["value"]
-    params[:challenge] = json["challenge"].reverse
-    params
+  describe '#activate_account' do
+    let(:token) { "asdfasdf" }
+    before do
+      UsersController.any_instance.stubs(:honeypot_or_challenge_fails?).returns(false)
+    end
+
+    context 'invalid token' do
+      it 'return success' do
+        put "/u/activate-account/#{token}"
+        expect(response).to be_success
+        expect(flash[:error]).to be_present
+      end
+    end
+
+    context 'valid token' do
+      let(:user) { Fabricate(:user) }
+
+      context 'welcome message' do
+        before do
+          EmailToken.expects(:confirm).with("#{token}").returns(user)
+        end
+
+        it 'enqueues a welcome message if the user object indicates so' do
+          user.send_welcome_message = true
+          user.expects(:enqueue_welcome_message).with('welcome_user')
+
+          put "/u/activate-account/#{token}"
+        end
+
+        it "doesn't enqueue the welcome message if the object returns false" do
+          user.send_welcome_message = false
+          user.expects(:enqueue_welcome_message).with('welcome_user').never
+
+          put "/u/activate-account/#{token}"
+        end
+      end
+
+      context "honeypot" do
+        it "raises an error if the honeypot is invalid" do
+          UsersController.any_instance.stubs(:honeypot_or_challenge_fails?).returns(true)
+          put "/u/activate-account/#{token}"
+          expect(response).not_to be_success
```

Comment:
Let us change this to test for the actual response code. Is it a `400` or `404`? Better to be explicit here.